analysis/src/inference.rs

```diff
@@ -0,0 +1,176 @@
+                    }
+                    (None, None) => {
+                        errors.push(format!("Impossible to infer the assignment register for `{a}`. Try being more explicit."));
+                    }
+                }
+            }
+        }
+    }
+
+    if !errors.is_empty() {
+        Err(errors)
+    } else {
+        Ok(machine)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ast::asm_analysis::AssignmentStatement;
+    use number::Bn254Field;
+
+    use crate::test_util::infer_str;
+
+    use super::*;
+
+    #[test]
+    fn inferred() {
+        let file = r#"
+            machine Incompatible {
+                reg pc[@pc];
+                reg X[<=];
+                reg Y[<=];
+                reg A;
+
+                instr foo -> X {}
+
+                function main {
+                    A <== foo();
+                }
+            }
+        "#;
+
+        let file = infer_str::<Bn254Field>(file).unwrap();
+
+        if let FunctionStatement::Assignment(AssignmentStatement { using_reg, .. }) =
+            &file.machines["Incompatible"].functions[0].body.statements[0]
+        {
+            assert_eq!(*using_reg, Some("X".to_string()));
+        } else {
+            panic!()
+        }
+    }
+
+    #[test]
+    fn compatible() {
+        let file = r#"
+            machine Incompatible {
+                reg pc[@pc];
+                reg X[<=];
+                reg Y[<=];
+                reg A;
+
+                instr foo -> X {}
+
+                function main {
+                    A <=X= foo();
+                }
+            }
+        "#;
+
+        let file = infer_str::<Bn254Field>(file).unwrap();
+
+        if let FunctionStatement::Assignment(AssignmentStatement { using_reg, .. }) =
+            &file.machines["Incompatible"].functions[0].body.statements[0]
+        {
+            assert_eq!(*using_reg, Some("X".to_string()));
+        } else {
+            panic!()
+        }
+    }
+
+    #[test]
+    fn incompatible() {
+        let file = r#"
+            machine Incompatible {
+                reg pc[@pc];
+                reg X[<=];
+                reg Y[<=];
+                reg A;
+
+                instr foo -> X {}
+
+                function main {
+                    A <=Y= foo();
+                }
+            }
+        "#;
+
+        assert_eq!(infer_str::<Bn254Field>(file).unwrap_err(), vec!["Assignment register `Y` is incompatible with `foo()`. Try replacing `<=Y=` by `<==`."]);
+    }
+
+    #[test]
+    fn unclear() {
+        let file = r#"
+            machine Incompatible {
+                reg pc[@pc];
+                reg X[<=];
+                reg Y[<=];
+                reg A;
+
+                function main {
+                    A <== 1;
+                }
+            }
+        "#;
+
+        assert_eq!(infer_str::<Bn254Field>(file).unwrap_err(), vec!["Impossible to infer the assignment register for `A <== 1;`. Try being more explicit.".to_string()]);
```

Comment:
Let's make an issue to also automate this by choosing a free assignment register

Reply:
https://github.com/powdr-labs/powdr/issues/454